src/boards/rainbow_final.cpp

```diff
@@ -0,0 +1,971 @@
+		uint32 freq3;
+		static uint8 b3 = 0;
+		static int32 phaseacc = 0;
+		static uint32 duff = 0;
+
+		freq3 = (vpsg2[1] + ((vpsg2[2] & 15) << 8) + 1);
+
+		for (V = start; V < end; V++) {
+			saw1phaseacc -= nesincsize;
+			if (saw1phaseacc <= 0)
+			{
+				int32 t;
+ rea:
+				t = freq3;
+				t <<= 18;
+				saw1phaseacc += t;
+				phaseacc += vpsg2[0] & 0x3f;
+				b3++;
+				if (b3 == 7)
+				{
+					b3 = 0;
+					phaseacc = 0;
+				}
+				if (saw1phaseacc <= 0)
+					goto rea;
+				duff = (((phaseacc >> 3) & 0x1f) << 4) * 6 / 8;
+			}
+			Wave[V >> 4] += duff;
+		}
+	}
+}
+
+static INLINE void DoSQVHQ(int x) {
+	int32 V;
+	int32 amp = ((vpsg1[x << 2] & 15) << 8) * 6 / 8;
+
+	if ((channels & (x + 1)) == 0)
+	{
+		cvbc[x] = 0;
+		return;
+	}
+
+	if (vpsg1[(x << 2) | 0x2] & 0x80)
+	{
+		if (vpsg1[x << 2] & 0x80)
+		{
+			for (V = cvbc[x]; V < (int)SOUNDTS; V++)
+				WaveHi[V] += amp;
+		}
+		else
+		{
+			int32 thresh = (vpsg1[x << 2] >> 4) & 7;
+			for (V = cvbc[x]; V < (int)SOUNDTS; V++) {
+				if (dcount[x] > thresh)
+					WaveHi[V] += amp;
+				vcount[x]--;
+				if (vcount[x] <= 0)
+				{
+					vcount[x] = (vpsg1[(x << 2) | 0x1] | ((vpsg1[(x << 2) | 0x2] & 15) << 8)) + 1;
+					dcount[x] = (dcount[x] + 1) & 15;
+				}
+			}
+		}
+	}
+	cvbc[x] = SOUNDTS;
+}
+
+static void DoSQV1HQ(void) {
+	DoSQVHQ(0);
+}
+
+static void DoSQV2HQ(void) {
+	DoSQVHQ(1);
+}
+
+static void DoSawVHQ(void) {
+	static uint8 b3 = 0;
+	static int32 phaseacc = 0;
+	int32 V;
+
+	if ((channels & 0x04) == 0)
+	{
+		cvbc[2] = 0;
+		return;
+	}
+
+	if (vpsg2[2] & 0x80)
+	{
+		for (V = cvbc[2]; V < (int)SOUNDTS; V++) {
+			WaveHi[V] += (((phaseacc >> 3) & 0x1f) << 8) * 6 / 8;
+			vcount[2]--;
+			if (vcount[2] <= 0)
+			{
+				vcount[2] = (vpsg2[1] + ((vpsg2[2] & 15) << 8) + 1) << 1;
+				phaseacc += vpsg2[0] & 0x3f;
+				b3++;
+				if (b3 == 7) {
+					b3 = 0;
+					phaseacc = 0;
+				}
+			}
+		}
+	}
+	cvbc[2] = SOUNDTS;
+}
+
+void RainbowFinalSound(int Count) {
+	int x;
+
+	DoSQV1();
+	DoSQV2();
+	DoSawV();
+	for (x = 0; x < 3; x++)
+		cvbc[x] = Count;
+}
+
+void RainbowFinalSoundHQ(void) {
+	DoSQV1HQ();
+	DoSQV2HQ();
+	DoSawVHQ();
+}
+
+void RainbowFinalSyncHQ(int32 ts) {
+	int x;
+	for (x = 0; x < 3; x++) cvbc[x] = ts;
+}
+
+static void RainbowFinalESI(void) {
+	GameExpSound.RChange = RainbowFinalESI;
+	GameExpSound.Fill = RainbowFinalSound;
+	GameExpSound.HiFill = RainbowFinalSoundHQ;
+	GameExpSound.HiSync = RainbowFinalSyncHQ;
+
+	memset(cvbc, 0, sizeof(cvbc));
+	memset(vcount, 0, sizeof(vcount));
+	memset(dcount, 0, sizeof(dcount));
+	if (FSettings.SndRate)
+	{
+		if (FSettings.soundq >= 1)
+		{
+			sfun[0] = DoSQV1HQ;
+			sfun[1] = DoSQV2HQ;
+			sfun[2] = DoSawVHQ;
+		} else {
+			sfun[0] = DoSQV1;
+			sfun[1] = DoSQV2;
+			sfun[2] = DoSawV;
+		}
+	}
+	else
+		memset(sfun, 0, sizeof(sfun));
+	AddExState(&SStateRegs, ~0, 0, 0);
+}
+
+/*
+// NSF Init
+
+void NSFRainbowFinal_Init(void) {
+	RainbowFinalESI();
+	SetWriteHandler(0x8000, 0xbfff, RainbowFinalSW);
+}
+*/
```

Comment:
Commented code. If there is a real reason to keep it, add an explanation why this code is deactivated. (Also `#if 0` is better than multilines comment for deactivating code (it allows multilines comments in deactivated code))

Reply:
I'd love to have NSF support later, and apparently that's all that it takes to enable it. However I don't have an NSF test file for it yet, that's why I commented those lines.
I'll `#if 0` and add a note.